drivers/SmartThings/jbl/src/discovery.lua

```diff
@@ -0,0 +1,104 @@
+local log = require "log"
+local discovery = {}
+
+local fields = require "fields"
+local discovery_mdns = require "discovery_mdns"
+
+local socket = require "cosock.socket"
+
+-- mapping from device DNI to info needed at discovery/init time
+local device_discovery_cache = {}
+
+local function set_device_field(driver, device)
+
+  log.info("set_device_field : dni = " .. tostring(device.device_network_id))
+  local device_cache_value = device_discovery_cache[device.device_network_id]
+
+  -- persistent fields
+  device:set_field(fields.DEVICE_IPV4, device_cache_value.ip, {persist = true})
+  device:set_field(fields.DEVICE_INFO, device_cache_value.device_info , {persist = true})
+  device:set_field(fields.CREDENTIAL, device_cache_value.credential , {persist = true})
+end
+
+local function update_device_discovery_cache(driver, dni, ip, credential)
+  log.info("update_device_discovery_cache for device dni: " .. tostring(dni) .. ", " .. tostring(ip))
+  local device_info = driver.discovery_helper.get_device_info(driver, dni, ip)
+  device_discovery_cache[dni] = {
+    ip = ip,
+    device_info = device_info,
+    credential = credential,
+  }
+end
+
+local function try_add_device(driver, device_dni, device_ip)
+  log.trace("try_add_device : dni=" .. tostring(device_dni) .. ", ip=" .. tostring(device_ip))
+
+  local credential  = driver.discovery_helper.get_credential(driver, device_dni, device_ip)
+
+  if not credential then
+    log.error("failed to get credential. dni=" .. device_dni .. ", ip=" .. device_ip)
+    return
+  end
+
+  update_device_discovery_cache(driver, device_dni, device_ip, credential)
+  local create_device_msg = driver.discovery_helper.get_device_create_msg(driver, device_dni, device_ip)
+  driver:try_create_device(create_device_msg)
+end
+
+function discovery.device_added(driver, device)
+  log.info("device_added : dni = " .. tostring(device.device_network_id))
+  set_device_field(driver, device)
+  device_discovery_cache[device.device_network_id] = nil
+  driver.lifecycle_handlers.init(driver, device)

```

Comment:
`init` is supposed to get called automatically after the added handler completes.

There *is* a bug where sometimes the `init` handler *doesn't* get called after `added`, and we are still root causing this; however, in our other land drivers, we use some extra state to make sure that we aren't calling the `added` and `init` handlers multiple times on a device.

If your `init` handler is idempotent, then it should be okay, but it's still doing extra work, and I would suggest adding similar checks like we have in our other drivers, to make sure you don't call `init` on a new device two times.

Reply:
Done, by checking device:get_field(fields._INIT)